include/envoy/network/io_handle.h

```diff
@@ -7,27 +7,93 @@
 namespace Envoy {
 namespace Network {
 
+namespace Address {
+class Instance;
+} // namespace Address
+
+/**
+ * Basic type for return result which has a return code and error code defined
+ * according to different implementation.
+ */
+template <typename T> struct IoHandleCallResult {
+  virtual ~IoHandleCallResult() {}
+
+  T rc_;
+  int errno_;
+};
+
+typedef IoHandleCallResult<int> IoHandleCallIntResult;
+typedef IoHandleCallResult<ssize_t> IoHandleCallSizeResult;
+
 /**
  * IoHandle: an abstract interface for all I/O operations
  */
 class IoHandle {
 public:
-  IoHandle() {}
-
   virtual ~IoHandle() {}
 
   /**
    * Return data associated with IoHandle.
    *
    * TODO(sbelair2) remove fd() method
+   * We probably still need some method similar to this one for
+   * evconnlistener_new(). Or We can move it to IoSocketHandle and down cast the
+   * IoHandle to IoSocketHandle wherever needed.
    */
   virtual int fd() const PURE;
 
   /**
    * Clean up IoHandle resources
    */
-  virtual void close() PURE;
+  virtual IoHandleCallIntResult close() PURE;
+
+  virtual bool isClosed() PURE;
+
+  virtual IoHandleCallSizeResult readv(const iovec* iovec, int num_iovec) PURE;
+
+  virtual IoHandleCallSizeResult writev(const iovec* iovec, int num_iovec) PURE;
+
+  virtual IoHandleCallIntResult bind(const Network::Address::Instance& address) PURE;
+
+  virtual IoHandleCallIntResult connect(const Network::Address::Instance& server_address) PURE;
+
+  /**
+   * Wrap setsockopt()
+   */
+  virtual IoHandleCallIntResult setIoHandleOption(int level, int optname, const void* optval,
+                                                  socklen_t optlen) PURE;
+  /**
+   * Wrap getsockopt()
+   */
+  virtual IoHandleCallIntResult getIoHandleOption(int level, int optname, void* optval,
+                                                  socklen_t* optlen) PURE;
+
+  /**
+   * Wrap getsockname()
+   */
+  virtual IoHandleCallIntResult getIoHandleName(const Network::Address::Instance& address) PURE;
+
+  virtual IoHandleCallIntResult getPeerName(const Network::Address::Instance& address) PURE;
+
+  /**
+   * Wrap fcntl(fd_, F_SETFL...)
+   */
+  virtual IoHandleCallIntResult setIoHandleFlag(int flag) PURE;
+  /**
+   * Wrap fcntl(fd_, F_GETFL...)
+   */
+  virtual IoHandleCallIntResult getIoHandleFlag() PURE;
+
+  virtual IoHandleCallIntResult listen(int backlog) PURE;
+
+  /**
+   * Wrap dup()
+   */
+  virtual std::unique_ptr<IoHandle> dup() PURE;
+
+  virtual IoHandleCallIntResult shutdown(int how) PURE;
```

Comment:
enum for shutdown type?

Reply:
I agree -- the values are different on Windows + POSIX (SD_RECEIVE, SD_SEND, SD_BOTH vs. SHUT_RD, SHUT_WR, SHUT_RDWR)